qualcommax: ipq60xx: add 8Devices Mango support

mantas-p opened the pull request:
Add ipq60xx subtarget. This PR is based on work done by @robimarko at [(1)](https://github.com/robimarko/openwrt/commits/qualcommax-ipq60xx/) with Linksys MR7350, though I had to remove device specific bits as support was incomplete (BDF, etc.) since this PR also adds WiFi support. 

Currently only WiFi firwmare that is stable is v2.4. Publicly available v2.7 crashes on ifconfig down.

SFP slot support is added as device variant, since runtime detection would not be possible with all revisions of Mango DVK.

Known issues include (https://github.com/openwrt/openwrt/issues/12891) warning on sysupgrade, but this is common issue with JFFS2 devices on linux-6.1.

Related BDF PR: https://github.com/openwrt/firmware_qca-wireless/pull/26

mantas-p committed:
ath11k-firmware: add wifi firmware for ipq6018

So far only stable FW is v2.4. It exists in a different
git repo, so add custom download routine.

Signed-off-by: Mantas Pucka <mantas@8devices.com>

mantas-p committed:
ipq-wifi: add support for 8devices Mango board

Add support for .ipq6018 BDF suffix and add Mango board entry

Signed-off-by: Mantas Pucka <mantas@8devices.com>

mantas-p committed:
mac80211: ath11k: disable coldboot for ipq6018

Coldboot calibration does not work at the moment and causes failure during
wifi startup.

Signed-off-by: Mantas Pucka <mantas@8devices.com>

robimarko commented on package/kernel/mac80211/patches/ath11k/906-ath11k-disable-coldboot-calibration-for-ipq6018.patch:
```diff
@@ -0,0 +1,13 @@
+--- a/drivers/net/wireless/ath/ath11k/core.c
```
Please export this with `git format-patch` with a proper title, description, and SoB.

robimarko commented:
For the sub-target commit itself, can you please remove all of the placeholder scripts for u-boot env, networking, etc?
Those should be added by board commits that actually need them.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk-sfp.dts:
```diff
@@ -0,0 +1,22 @@
+};
+
+&switch {
+	switch_mac_mode1 = <0x14>;
```
Please use dt-binding include values and not magic values.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+	pinctrl-names = "default";
+	status = "okay";
+
+	m25p80@0 {
```
`flash@0`

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
```
Is the SMEM partition table correct?
If so, please use that

robimarko replied:
Sorry, just noticed that there is NAND as well so it cannot be used.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition0 {
```
`partition@0`

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+				reg = <0x00000000 0x000c0000>;
+			};
+
+			partition1 {
```
`partition@c0000`

And so on for the reset

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+		pinctrl-0 = <&led_pins>;
+		pinctrl-names = "default";
+
+		led@66 {
```
`wlan2g`

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+			linux,default-trigger = "phy0radio";
+		};
+
+		led@67 {
```
`wlan2g`

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+			label = "green:wlan5g";
+			gpios = <&tlmm 66 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "phy0radio";
+		};
```
Please add the color and function properties as we will be moving to those rather soon instead of label

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+&dp3 {
+	status = "okay";
+
+	qcom,id = <4>;
```
Why are you overriding the default?

mantas-p replied:
IIRC this was needed in QSDK for QCA8072/8075 compatibility, but I'll recheck if it's really needed here.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+&dp4 {
+	status = "okay";
+
+	qcom,id = <3>;
```
Why are you overriding the default?

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+
+	vqmmc-supply = <&ipq6018_l2>;
+	cd-gpios = <&tlmm 62 GPIO_ACTIVE_LOW>;
+	disable-wp;
```
Since there is no WP GPIO described, you don't need this anyway.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+	switch_lan_bmp = <(ESS_PORT3 | ESS_PORT3)>;
+	switch_wan_bmp = <ESS_PORT5>;
+	switch_mac_mode = <MAC_MODE_PSGMII>;
+	switch_mac_mode1 = <MAC_MODE_SGMII_CHANNEL0>;
```
Why not `MAC_MODE_SGMII_PLUS` ?

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+	switch_mac_mode1 = <MAC_MODE_SGMII_CHANNEL0>;
+
+	qcom,port_phyinfo {
+		port@3 {
```
`port@4`

And so on

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition0 {
```
`partition@0`

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+			#size-cells = <1>;
+
+			partition0 {
+				label = "nand_data";
```
What is this used for?

mantas-p replied:
NAND chip is shipped empty by default. Partition is for users to be able to access available storage.

robimarko replied:
OK, that makes sense.

robimarko commented on target/linux/qualcommax/ipq60xx/base-files/etc/board.d/01_leds:
```diff
@@ -6,6 +6,11 @@ board_config_update
 case "$board" in
+8devices,mango-dvk|\
+8devices,mango-dvk-sfp)
+	ucidef_set_led_wlan "wlan5g" "WIFI 5GHz" "green:wlan5g" "phy0radio"
```
There should be no reason for userspace trigger setting, you can set these via DT

jokubasver commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-mango-dvk.dts:
```diff
@@ -0,0 +1,347 @@
+&switch {
+	status = "okay";
+
+	switch_lan_bmp = <(ESS_PORT3 | ESS_PORT3)>;
```
`switch_lan_bmp = <(ESS_PORT3 | ESS_PORT4)>;`

Ansuel commented:
@mantas-p what problem did you find with the 2.7 fw aside from wifi down?

From what I notice sometimes it does crash with
```
[  885.421771] QC Image Version: QC_IMAGE_VERSION_STRING=WLAN.HK.2.7.0.1-01744-QCAHKSWPL_SILICONZ-1
[  885.421771] Image Variant : IMAGE_VARIANT_STRING=6018.wlanfw.evalQ
[  885.421771]
[  885.421771] NOC_error.c:474 NOCError: FATAL ERRORparam0 :zero, param1 :zero, param2 :zero.
[  885.421771] Thread ID      : 0x00000070  Thread name    : ABTIST182  Process ID     : 0
[  885.421771] Register:
[  885.421771] SP : 0x4af83428
[  885.421771] FP : 0x4af83430
[  885.421771] PC : 0x4ac081b0
[  885.421771] SSR : 0x00000008
[  885.421771] BADVA : 0x00020000
[  885.421771] LR : 0x4ac0794c
[  885.421771]
[  885.421771] Stack Dump
[  885.421771] from : 0x4af83428
[  885.421771] to   : 0x4af83500
[  885.421771]
[  885.468774] remoteproc remoteproc0: crash detected in cd00000.remoteproc: type fatal error
[  885.490901] remoteproc remoteproc0: handling crash #1 in cd00000.remoteproc
[  885.498952] remoteproc remoteproc0: recovering cd00000.remoteproc
[  885.540992] remoteproc remoteproc0: stopped remote processor cd00000.remoteproc
[  885.835491] remoteproc remoteproc0: remote processor cd00000.remoteproc is now up
```
and as you said on wifi down with
```
[  538.138134] br-lan: port 6(phy0-ap0) entered disabled state
[  538.722472] qcom-q6v5-wcss-pil cd00000.remoteproc: fatal error received:
[  538.722472] QC Image Version: QC_IMAGE_VERSION_STRING=WLAN.HK.2.7.0.1-01744-QCAHKSWPL_SILICONZ-1
[  538.722472] Image Variant : IMAGE_VARIANT_STRING=6018.wlanfw.evalQ
[  538.722472]
[  538.722472]     :Excep  :0 Exception detectedparam0 :zero, param1 :zero, param2 :zero.
[  538.722472] Thread ID      : 0x00000060  Thread name    : WLAN RT1  Process ID     : 0
[  538.722472] Register:
[  538.722472] SP : 0x4baa23a0
[  538.722472] FP : 0x4baa23b0
[  538.722472] PC : 0x4acdd350
[  538.722472] SSR : 0x00000001
[  538.722472] BADVA : 0x00000387
[  538.722472] LR : 0x4acdd358
[  538.722472]
[  538.722472] Stack Dump
[  538.722472] from : 0x4baa23a0
[  538.722472] to   : 0x4baa2908
[  538.722472]
[  538.768677] remoteproc remoteproc0: crash detected in cd00000.remoteproc: type fatal error
[  538.790894] remoteproc remoteproc0: handling crash #1 in cd00000.remoteproc
[  538.799220] remoteproc remoteproc0: recovering cd00000.remoteproc
[  538.842638] remoteproc remoteproc0: stopped remote processor cd00000.remoteproc
[  538.842782] ath11k c000000.wifi: failed to clear rx_filter for monitor status ring: (-108)
[  539.098577] br-lan: port 5(phy1-ap0) entered disabled state
[  539.098950] device phy1-ap0 left promiscuous mode
[  539.103025] br-lan: port 5(phy1-ap0) entered disabled state
[  539.137281] remoteproc remoteproc0: remote processor cd00000.remoteproc is now up
```

but from what I can see it does correctly recover. Did you observe other big problem with this?

mantas-p commented:
@Ansuel I've seen wifi down issue only, but for me it does not seem to recover correctly - after crash even if beacons are transmitted I couldn't connect anyway. Only after `rmmod ath11k_ahb; modprobe ath11k_ahb; wifi` it seems to recover fully.

mantas-p committed:
qualcommax: ipq60xx: Add 8devices Mango DVK

8devices Mango DVK is a single board computer / devkit for 8devices Mango
system-on-module (SoM).

Specifications:
* CPU: Qualcomm IPQ6010 Quad core Cortex-A53 1.8GHz
* RAM: 512 MB
* Storage:
    * 32 MB serial NOR flash (on SoM)
    * 256 MB parallel NAND flash (on DVK)
* Ethernet:
    * 2x1G RJ45 ports(QCA8072 or QCA8075)
    * 1x2.5G RJ45 port (QCA8081)
    * 1xSFP (shares SGMII with QCA8081)
* Switch: Qualcomm Atheros IPQ6010
* WLAN:
    * 2.4GHz: QCN5121 2x2 802.11b/g/n/ax 574 Mbps PHY rate
    * 5GHz: QCN5152 2x2 802.11a/n/ac/ax 1201 Mbps PHY rate
* USB:
    * 1x USB3.0 Type-A port
    * 1x USB2.0 available at mini PCIe slot
* PCIe: 1x mini PCIe slot 1xLane Gen3 (8GT/s)
* SD/eMMC (on a single shared bus - only one can be active):
    * micro SD slot
    * eMMC module connector
* LEDs:
    * Green power led (not controllable)
    * Green 2.4GHz radio led (GPIO 67)
    * Green 5GHz radio led (GPIO 66)
* Buttons:
    * 1x (WPS GPIO79) button
* GPIOs: 2.54mm header brings out 18 GPIOs (1.8V level)
* UART: 4-pin UART header (3.3V level)
    * 115200 8N1, 3.3V-Tx-Rx-GND (3.3V is pin 1 close to boot-switch SW2)
* Power:
    * PoE IN on 2.5G port (passive 24-48V)
    * DC power terminal (12-58V)

Installation instructions:

Vendor image format is compatible with squashfs-sysupgrade image. Run:

sysupgrade -n -F openwrt-qualcommax-ipq60xx-8devices_mango-dvk-squashfs-sysupgrade.bin

Signed-off-by: Mantas Pucka <mantas@8devices.com>

mantas-p commented:
Thanks everyone for reviews. I've updated PR with these main changes:
* Rebase
* Changes to address reviews
* Add patch, that I initially missed, to fix QCA8075 on Mango: `package/kernel/qca-ssdk/patches/0101-qca-ssdk-use-generic-way-to-map-PSGMII-ports.patch`

robimarko commented:
@mantas-p May I ask what the backstory in regards to the newly introduced PSGMII remapping patch is?

mantas-p commented:
PR update:
* Rebase
* Add thermal sensor support:
```
target/linux/qualcommax/patches-6.1/0140-arm64-dts-qcom-ipq6018-add-tsens-node.patch
target/linux/qualcommax/patches-6.1/0141-arm64-dts-qcom-ipq6018-add-thermal-zones.patch
```

mantas-p commented:
PR update:
  * Rebase
  * Enable APM/CPR4 driver
  * Update upstreamed patches
  * Enable QDSS_AT clock in remoteproc driver since upstreamed clock patch dropped CLK_IS_CRITICAL flag

Ansuel commented:
@mantas-p I cherry picked some of the commits here to move thins further.

mantas-p commented:
@Ansuel Thanks for pulling the target in!

I've considered net/phy/sfp.c driver for SFP support, but that requires I2C bus which is not routed to SFP cage on this board.

Not sure if new PHY driver remove need for PSGMII remapping, as it is done on the SoC side. I guess I'll revisit  this once new qca8075 driver gets merged.

mantas-p committed:
qca-ssdk: support selecting PCS channel for PORT3 on IPQ6018

When QCA8072 is used in PSGMII mode with IPQ6018, PCS used for second
PHY port would overlap with one used by SGMII+ port. SoC has register
to select different PCS in such case.

Original code used PHY_ID for this decision, which also had other
issues, but is no longer viable since we moved to upstream QCA807x
driver.

Introduce DT property port3_pcs_channel to allow describing this in DT.
Default value is <2>, and for some QCA8072 designs <4> would be needed.

Signed-off-by: Mantas Pucka <mantas@8devices.com>

mantas-p committed:
qualcommax: ipq60xx: set correct PHY mode for port 0-4

[port 8ed390a (qualcommax: set correct PHY mode for port 0-4) to ipq60xx]

Port 0-4 have the mode set to SGMII instead of PSGMII. Now that we use
he upstream qca807x driver, this conflicts with the qca SSDK driver
that expects the mode to be PSGMII as for not integrated driver, it does
refer to the real PHY mode.

Update the entry for port 0-4 to PSGMII to solve warning from qca SSDK
in ipq6018-ess.dtsi.

Signed-off-by: Mantas Pucka <mantas@8devices.com>

robimarko commented:
BDF has been merged, you can include the ipq-wifi update

mantas-p committed:
ipq-wifi: update to Git HEAD (2024-02-19)

1cc59e1 ipq6018: add 8devices Mango BDF
10279cc ipq40xx:Add support for Netgear LBR20 with two BDF's

Signed-off-by: Mantas Pucka <mantas@8devices.com>

robimarko commented:
Was time to get this finally merged, I would really love to see WLAN fixed up with newer FW though
